File: services/tracking/src/features/tracking/commands/update_status.py
"""`PUT /v1/trackings/{order_id}/status` command handler (JE-94).

The carrier webhook's write path. Transport-free like `create_tracking`: it takes a
plain dataclass and returns the persisted entity, so the FastAPI handler is a thin
translation and this logic is testable without a server.

## Why the lookup here is UNSCOPED

The caller is an external carrier authenticated by an API key, not a user behind a
Cognito JWT — its gateway route carries no authorizer, so there is no
`x-user-id` on the request at all and no identity to scope by. The repository is
therefore called with `user_id=None`, the same unscoped mode the gRPC reads use.

This is the single most confusable thing in Phase D: the two REST *reads* sitting
beside this command DO scope by `user_id`, and reusing their filter here would make
every carrier update return 404 — the endpoint would look implemented and never
work. The design calls this out explicitly ("An implementer who assumes
`x-user-id` is present here will write broken code").

## Why the guards are applied here and not in the repository

`assert_can_transition` lives in `domain/status.py` as a pure function, and
`TrackingRepository.update_status` deliberately does not validate. This layer is
where the decision is made, so the same guards stay reusable by TestMode
progression (Phase E), which has no request and no HTTP status code to map to.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

from sqlalchemy.orm import Session

from src.features.tracking.domain.models import Tracking
from src.features.tracking.domain.repository import TrackingRepository
from src.features.tracking.domain.status import (
    TrackingStatus,
    assert_can_transition,
    parse_status,
)
from src.shared.audit.audit_actor import AuditActor
from src.shared.messaging.event_publisher import EventPublisher

logger = logging.getLogger(__name__)


class TrackingNotFoundError(Exception):
    """No live tracking exists for the given `order_id`.

    Distinct from a rejected transition: the handler maps this to `404` and
    `InvalidTransitionError` to `400`. Note there is no ownership dimension here —
    unlike the reads, a 404 from this endpoint means the order genuinely has no
    tracking, because nothing was filtered out by identity.
    """

    def __init__(self, order_id: str) -> None:
        self.order_id = order_id
        super().__init__(f"no tracking for order_id {order_id}")


@dataclass(frozen=True, slots=True)
class UpdateTrackingStatusCommand:
    """Input to the carrier status update.

    Carries no `user_id` — there is none on this request. See the module docstring.
    """

    order_id: str
    #: The raw string off the request body; parsed (and rejected) below.
    status: str


def update_tracking_status(
    session: Session,
    command: UpdateTrackingStatusCommand,
    *,
    actor: AuditActor = AuditActor.CARRIER_STATUS_UPDATE,
    publisher: EventPublisher | None = None,
) -> Tracking:
    """Advance a tracking to `command.status`, appending a history row.

    `actor` is the ONLY thing that differs between this function's two callers.
    The carrier PUT takes the default; TestMode progression (Phase E) passes
    `AuditActor.TEST_MODE_PROGRESSION` so an automatic run stays identifiable from
    `tracking_history.created_by` after the fact. Everything else — the parse, the
    lookup, the guards, the persistence — is deliberately shared: a second
    implementation for the automatic path is how the two would start disagreeing
    about what a transition means.

    Order of operations is load-bearing:

    1. **Parse** the status. An unknown value raises `ValueError` before anything
       is read, so garbage never reaches the state machine.
    2. **Find** the tracking, unscoped by `order_id` alone. Missing → 404.
    3. **Guard** the transition. A rejection raises `InvalidTransitionError`
       carrying its machine-readable reason → 400, and nothing is written.
    4. **Persist**, which updates the status and appends the transition in one unit
       of work. The caller's `write_session` owns the commit, so the status change
       and its history row land together or not at all.

    Steps 2 and 3 are separate so a rejected transition on an existing tracking
    cannot be confused with a missing one — different status codes, different
    causes.

    ## 5. Emit `TRACKING_STATUS_CHANGED` (events-pipeline milestone)

    Every successful transition publishes one event, `DELIVERED` included —
    there is no suppression, so a TestMode run produces four: creation writes
    `PLACED` without emitting, and the four steps that follow each emit one.

    Emission lives HERE, and only here, for the same reason the guards do: this
    is the single write path behind BOTH the carrier PUT and TestMode
    progression, so one call site covers both callers. A second emission point
    beside either of them is how the two would start disagreeing about which
    transitions notify.

    **The event is built from the PERSISTED ENTITY, and could not come from
    anywhere else.** This request has no caller identity at all (see "Why the
    lookup here is UNSCOPED" above): the carrier presents an API key, its
    gateway route has no Cognito authorizer, and no `x-user-id` reaches the
    service. So `updated` — the row this function just wrote — is the only
    source of an owner for the event, and it is handed to the publisher whole
    rather than decomposed into keyword arguments here.

    Passing the entity is what lets the payload carry the shipment's own data:
    the tracking number, the delivery address and the FULL transition history
    that makes the email's five-step timeline renderable. `Tracking.history` is
    `lazy="selectin"` and ordered by the relationship, so the publisher neither
    issues an N+1 nor re-derives an ordering that is already correct. Note that
    `repository.update_status` EXPIRES that collection after appending, so the
    publisher's read of it reloads — which is the point: the timeline it
    serializes contains the transition being announced, rather than the stale
    pre-update list.

    `cognito_sub` is deliberately not the envelope's `user_id` — that is the
    ownership key the REST reads filter by, and Users is keyed by the internal
    `usr_` id, so a sub would resolve to no email and the notification would be
    dropped as `no_email_for_user`.

    **The envelope's `author` is this function's `actor`, and carries no human.**
    `user_id` above says who the event is ABOUT; `author` says what ORIGINATED
    it, and on both of this command's paths that is a system: an external carrier
    or a TestMode timer. So the author's own `user_id`/`cognito_sub` are omitted
    rather than backfilled with the order's owner — attributing a carrier's
    status update to the buyer would be plainly false, and it is the reason the
    two fields are separate. `actor` is passed down (never fixed in the
    publisher) so the two paths stay distinguishable on the wire exactly as they
    already are in `tracking_history.created_by`.

    `publisher` defaults to the real shared SQS publisher, so the carrier and
    TestMode paths need no call-site change; tests inject a recording fake, or
    `NoopEventPublisher` when they must not emit. The default is resolved
    LAZILY, inside the call rather than in the signature, so importing this
    module never constructs a boto3 client nor requires a valid environment.

    Publishing cannot fail this command: `SqsEventPublisher` logs and swallows
    its own failures (the reasoning is in its class docstring — the transition is
    already committed, and a 500 would make the carrier retry a status change we
    actually recorded, which the forward-only guard then rejects as a 400).
    """
    requested: TrackingStatus = parse_status(command.status)

    repository = TrackingRepository(session)
    # user_id is NOT passed: unscoped by design — see the module docstring.
    tracking = repository.get_by_order_id(command.order_id)
    if tracking is None:
        raise TrackingNotFoundError(command.order_id)

    current = parse_status(tracking.status)
    assert_can_transition(current, requested)

    updated = repository.update_status(
        tracking=tracking,
        status=requested,
        actor=actor,
    )

    _emit_status_changed(
        publisher,
        # The PERSISTED row this function just wrote — never a request value.
        # Everything the event says about the shipment (its owner, its order,
        # its number, its address, its whole history) is read off this entity by
        # the publisher; see the docstring's section 5.
        tracking=updated,
        # The one fact the entity no longer holds: `updated.status` is already
        # the NEW status.
        previous_status=current.value,
        # THIS function's `actor` — the one already distinguishing the carrier
        # PUT from TestMode progression — travels onto the envelope as its
        # author. Passed down rather than fixed in the publisher, or a TestMode
        # run would be published as a carrier update; see the docstring's
        # section 5.
        actor=actor,
    )

    return updated


def _emit_status_changed(
    publisher: EventPublisher | None,
    *,
    tracking: Tracking,
    previous_status: str,
    actor: AuditActor,
) -> None:
    """Publish the transition, and never let doing so break the transition.

    ## Why building the publisher is inside the try, not just publishing

    `shared_event_publisher()` calls `get_settings()`, which constructs a
    pydantic `Settings` and raises `ValidationError` on an incomplete
    environment. **`ValidationError` is a subclass of `ValueError`** — and the
    carrier router catches `ValueError` to mean "`parse_status` rejected the
    status", mapping it to `400 invalid_status`.

    So a misconfigured environment (or a test app built without a full one, which
    is exactly how this suite builds its app) would surface as *"the carrier sent
    an invalid status"*: a 400 blaming the caller for a transition that was
    already written to the database, with a `reason` naming the wrong cause
    entirely. That is a genuinely misleading failure, and it is not
    hypothetical — it is what this function was written in response to.

    Catching here rather than widening the router's `except` keeps the rule where
    it belongs: publishing is a secondary effect, so NOTHING about it — not the
    send, not the client construction, not reading settings — may change the
    outcome of a transition that already succeeded.

    `SqsEventPublisher` swallows its own send/resolution failures (see its class
    docstring). This guard covers the layer beneath that: obtaining the publisher
    at all.
    """
    try:
        if publisher is None:
            # Imported here, not at module scope: the import itself pulls in
            # boto3 and the call reads settings, so a test injecting a fake
            # publisher touches neither.
            from src.shared.messaging.sqs_event_publisher import (
                shared_event_publisher,
            )

            publisher = shared_event_publisher()

        publisher.publish_tracking_status_changed(
            tracking=tracking,
            previous_status=previous_status,
            actor=actor,
        )
    except Exception:  # noqa: BLE001 - a notification must not fail a write
        logger.exception(
            "tracking_status_changed_publish_failed",
            extra={
                "app_event": "tracking_status_changed_publish_failed",
                "reason": "publisher_unavailable",
                # Read off the entity, like everything else on this path. These
                # three fields and no more: the line names WHICH notification
                # was lost, never what it would have contained — no address, no
                # name, no history. See [[logging-context]].
                "order_id": tracking.order_id,
                "user_id": tracking.user_id,
                "status": tracking.status,
            },
        )

File: services/tracking-go/internal/app/update_status.go
package app

import (
	"context"
	"time"

	"github.com/jemartinez/3mrai/services/tracking-go/internal/domain"
	"github.com/jemartinez/3mrai/services/tracking-go/internal/domain/audit"
)

// StatusWriter is this use case's own narrow port, declared here by its consumer.
//
// CONTRACT: GetByOrderID is UNSCOPED and a DIFFERENT method from the reads'
// GetByOrderIDScoped, never one method with an empty scope argument. This
// endpoint has no caller identity — the gateway route carries no authorizer — so
// the ownership filter would 404 every carrier call. And Go's zero string is "",
// so an optional argument silently means "scoped to the empty string".
// See [[user-id-vs-cognito-sub-ownership-key]]
type StatusWriter interface {
	GetByOrderID(ctx context.Context, orderID string) (domain.Tracking, error)

	// CONTRACT: ApplyTransition updates the parent, appends the history row and
	// RE-READS the history in one transaction, stamped from `now`. The re-read
	// is this method's job, not the caller's — a caller holding a pre-append
	// slice cannot know it is stale, and the returned value is both the 200 body
	// and the published event.
	ApplyTransition(ctx context.Context, t domain.Tracking, to domain.Status, actor audit.Actor, now time.Time) (domain.TrackingWithHistory, error)
}

// CONTRACT: EventPublisher returns no error, so a notification failure cannot
// fail a committed write. The actor is a PARAMETER, never a constant inside the
// publisher — a hardcoded one relabels automatic progressions as carrier
// updates. See [[audit-fields]]
type EventPublisher interface {
	PublishTrackingStatusChanged(ctx context.Context, t domain.TrackingWithHistory, previousStatus string, actor audit.Actor)
}

// CacheInvalidator clears a tracking's cache entries. Also returns nothing: a
// cache is an optimization, and an optimization that can fail a request is a
// liability.
type CacheInvalidator interface {
	InvalidateTracking(ctx context.Context, orderID, cognitoSub, userID string)
}

// UpdateStatus is the SINGLE write path behind BOTH the carrier PUT and TestMode
// progression. The ONLY thing that differs between its two callers is the actor.
type UpdateStatus struct {
	writer      StatusWriter
	publisher   EventPublisher
	invalidator CacheInvalidator
	clock       func() time.Time
}

// NewUpdateStatus wires the transition. A nil clock takes the production one.
func NewUpdateStatus(
	writer StatusWriter,
	publisher EventPublisher,
	invalidator CacheInvalidator,
	clock func() time.Time,
) *UpdateStatus {
	if clock == nil {
		// UTC and TRUNCATED to whole seconds. The DATETIME columns here have
		// fsp 0 and MySQL ROUNDS fractional seconds rather than truncating them,
		// so an untruncated 15:00:00.7 persists as 15:00:01 — one second ahead of
		// the value the response and the event report.
		clock = func() time.Time { return time.Now().UTC().Truncate(time.Second) }
	}
	return &UpdateStatus{
		writer:      writer,
		publisher:   publisher,
		invalidator: invalidator,
		clock:       clock,
	}
}

// Execute advances a tracking to `requested`, appending the transition to its
// history. `actor` is the ONLY difference between its two callers; everything
// else is shared so the two cannot disagree on what a transition means.
//
// CONTRACT: The step order is load-bearing. 1) Find UNSCOPED by order_id.
// 2) Guard; a rejection writes NOTHING and carries its reason. 3) Persist parent
// and history in one unit of work, then RE-READ. 4) Only after the commit,
// publish then invalidate. Steps 1 and 2 stay separate so a rejection is never
// confused with a missing tracking. See [[tracking-service-design]]
func (uc *UpdateStatus) Execute(
	ctx context.Context, orderID string, requested domain.Status, actor audit.Actor,
) (domain.TrackingWithHistory, error) {
	if actor == "" {
		actor = audit.CarrierStatusUpdate
	}

	tracking, err := uc.writer.GetByOrderID(ctx, orderID)
	if err != nil {
		return domain.TrackingWithHistory{}, err
	}

	// Captured BEFORE the write: the row's status is already the new one by the
	// time the event is built, so previous_status cannot be read off the entity.
	previous := tracking.Status

	if err := domain.AssertCanTransition(previous, requested); err != nil {
		return domain.TrackingWithHistory{}, err
	}

	now := uc.clock()

	// ApplyTransition commits, and the re-read of the history happens INSIDE it —
	// so the returned slice contains the transition being announced rather than
	// the stale pre-update list. The Python expires that collection for exactly
	// this reason: the carrier PUT once reported the NEW status alongside a
	// history that did not contain it.
	updated, err := uc.writer.ApplyTransition(ctx, tracking, requested, actor, now)
	if err != nil {
		return domain.TrackingWithHistory{}, err
	}

	// CONTRACT: Both of these run AFTER the commit and cannot fail the request.
	// Invalidating first opens a window where a concurrent read misses, sees the
	// pre-update row, and writes that stale body back under the key just
	// cleared — serving a superseded status for a full TTL. Identities come off
	// the PERSISTED ROW; the carrier sends none. See [[x-cache-response-header]]
	uc.publish(ctx, updated, string(previous), actor)
	uc.invalidate(ctx, updated)

	return updated, nil
}

// ContinueDeletedTestMode publishes one remaining TestMode transition after E2E
// cleanup has soft-deleted the tracking mid-run.
//
// CONTRACT: This is NOT a persistence path — the tombstone stays untouched and
// no live history row is recreated. It advances the last committed snapshot the
// progression captured, keeping the event chain complete for the fixture while
// soft-delete semantics hold. See [[soft-delete]]
func (uc *UpdateStatus) ContinueDeletedTestMode(
	ctx context.Context,
	current domain.TrackingWithHistory,
	requested domain.Status,
) (domain.TrackingWithHistory, error) {
	previous := current.Tracking.Status
	if err := domain.AssertCanTransition(previous, requested); err != nil {
		return domain.TrackingWithHistory{}, err
	}

	now := uc.clock()
	updatedTracking := current.Tracking
	updatedTracking.Status = requested
	updatedTracking.Datetime = now
	updatedTracking.UpdatedBy = string(audit.TestModeProgression)
	updatedTracking.UpdatedAt = now

	history := make([]domain.TrackingHistory, 0, len(current.History)+1)
	history = append(history, current.History...)
	history = append(history, domain.TrackingHistory{
		TrackingID: current.Tracking.ID,
		Status:     requested,
		UserID:     current.Tracking.UserID,
		OrderID:    current.Tracking.OrderID,
		CognitoSub: current.Tracking.CognitoSub,
		Datetime:   now,
		CreatedBy:  string(audit.TestModeProgression),
		CreatedAt:  now,
		UpdatedBy:  string(audit.TestModeProgression),
		UpdatedAt:  now,
	})
	updatedTracking.History = history
	updated := domain.TrackingWithHistory{
		Tracking: updatedTracking,
		History:  history,
	}

	uc.publish(ctx, updated, string(previous), audit.TestModeProgression)
	return updated, nil
}

// publish is best-effort and swallows everything, panics included.
//
// CONTRACT: A notification must not break the committed write it describes. A
// 500 makes the carrier retry a change we recorded, which the forward-only
// guard rejects as 400 — a permanent-looking failure for something that
// succeeded. See [[events-pipeline-design]]
func (uc *UpdateStatus) publish(
	ctx context.Context, t domain.TrackingWithHistory, previous string, actor audit.Actor,
) {
	defer func() {
		// Deliberately swallowed, not re-raised and not logged here: the
		// publisher owns its own failure logging with a machine-readable reason,
		// and this use case has no logger by design (it is transport-free).
		_ = recover()
	}()
	if uc.publisher == nil {
		return
	}
	uc.publisher.PublishTrackingStatusChanged(ctx, t, previous, actor)
}

// invalidate is the second best-effort step, and it is DELIBERATELY separate
// from publish rather than sharing one recover: a publisher panic must not skip
// the invalidation that follows it, or a Users outage would leave the cache
// serving the superseded status for a full TTL.
func (uc *UpdateStatus) invalidate(ctx context.Context, t domain.TrackingWithHistory) {
	defer func() { _ = recover() }()
	if uc.invalidator == nil {
		return
	}
	uc.invalidator.InvalidateTracking(ctx,
		t.Tracking.OrderID, t.Tracking.CognitoSub, t.Tracking.UserID)
}
